plugins/logitech-scribe/fu-logitech-scribe-device.c

```diff
@@ -0,0 +1,603 @@
+					ep,
+					(guint8 *)buf->data,
+					buf->len,
+					&transferred,
+					BULK_TRANSFER_TIMEOUT,
+					cancellable,
+					error)) {
+		g_prefix_error(error, "failed to send using bulk transfer: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_scribe_device_recv(FuLogitechScribeDevice *self,
+			       GByteArray *buf,
+			       gint interface_id,
+			       guint timeout,
+			       GError **error)
+{
+	gsize received_length = 0;
+	gint ep;
+	g_return_val_if_fail(buf != NULL, FALSE);
+
+	if (interface_id == BULK_INTERFACE_UPD) {
+		ep = self->update_ep[EP_IN];
+	} else {
+		g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_FAILED, "interface is invalid");
+		return FALSE;
+	}
+	if (!g_usb_device_bulk_transfer(fu_usb_device_get_dev(FU_USB_DEVICE(self->usb_device)),
+					ep,
+					buf->data,
+					buf->len,
+					&received_length,
+					timeout,
+					NULL,
+					error)) {
+		g_prefix_error(error, "failed to receive using bulk transfer: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_scribe_device_send_upd_cmd(FuLogitechScribeDevice *self,
+				       guint32 cmd,
+				       GByteArray *buf,
+				       GError **error)
+{
+	guint32 cmd_tmp = 0x0;
+	guint timeout = BULK_TRANSFER_TIMEOUT;
+	g_autoptr(GByteArray) buf_pkt = g_byte_array_new();
+	g_autoptr(GByteArray) buf_ack = g_byte_array_new();
+
+	fu_byte_array_append_uint32(buf_pkt, cmd, G_LITTLE_ENDIAN); /* Type(T) : Command type */
+	fu_byte_array_append_uint32(buf_pkt,
+				    buf != NULL ? buf->len : 0,
+				    G_LITTLE_ENDIAN); /*Length(L) : Length of payload */
+	if (buf != NULL) {
+		g_byte_array_append(buf_pkt,
+				    buf->data,
+				    buf->len); /* Value(V) : Actual payload data */
+	}
+	if (!fu_logitech_scribe_device_send(self, buf_pkt, BULK_INTERFACE_UPD, error))
+		return FALSE;
+
+	/* receiving INIT ACK */
+	fu_byte_array_set_size(buf_ack, MAX_DATA_SIZE, 0x00);
+
+	/* extending the bulk transfer timeout value, as android device takes some time to
+	   calculate Hash and respond */
+	if (CMD_END_TRANSFER == cmd)
+		timeout = HASH_TIMEOUT;
+
+	if (!fu_logitech_scribe_device_recv(self, buf_ack, BULK_INTERFACE_UPD, timeout, error))
+		return FALSE;
+
+	if (!fu_memread_uint32_safe(buf_ack->data,
+				    buf_ack->len,
+				    COMMAND_OFFSET,
+				    &cmd_tmp,
+				    G_LITTLE_ENDIAN,
+				    error))
+		return FALSE;
+	if (cmd_tmp != CMD_ACK) {
+		g_set_error(error, G_IO_ERROR, G_IO_ERROR_FAILED, "not CMD_ACK, got %x", cmd);
+		return FALSE;
+	}
+	if (!fu_memread_uint32_safe(buf_ack->data,
+				    buf_ack->len,
+				    UPD_PACKET_HEADER_SIZE,
+				    &cmd_tmp,
+				    G_LITTLE_ENDIAN,
+				    error))
+		return FALSE;
+	if (cmd_tmp != cmd) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_FAILED,
+			    "invalid upd message received, expected %x, got %x",
+			    cmd,
+			    cmd_tmp);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gchar *
+fu_logitech_scribe_device_compute_hash(GBytes *data)
```

Comment:
I think you can use `g_compute_checksum_for_bytes()` rather than this.

Reply:
Thanks Richard. I am on it. Not familiar with g_usb_device_get_interface(), g_compute_checksum_for_bytes() APIs, so searching/testing.

Reply:
@hughsie  
g_compute_checksum_for_bytes() is not working. Perhaps g_checksum_get_digest()/g_base64_encode() are required here.
Here is the old/new code with log output:
	base64hash = fu_logitech_scribe_device_compute_hash(fw);
	g_debug("SRS TODO TEMP base64hash: %s", base64hash);
	base64hash_new = g_compute_checksum_for_bytes(G_CHECKSUM_MD5, fw);
	g_debug("SRS TODO TEMP base64hash_new: %s", base64hash_new);
	
	Waiting…                 [*************************************  ]05:03:22.355 FuPluginLogitechScribe SRS TODO TEMP base64hash: o+4aP7n2iWGZbjiNo5w05w==
05:03:22.595 FuPluginLogitechScribe SRS TODO TEMP base64hash_new: a3ee1a3fb9f68961996e388da39c34e7
<snip>
failed to write end transfer transfer packet: failed to receive using bulk transfer: transfer timed out

Reply:
Ahh! I didn't notice it was g_base64_encode'd -- my mistake. The previous code is fine, sorry.

Reply:
Thanks.